### ISSUE TITLE
Make Rest manager more decouple

### ISSUE THREAD

**Issue by H01001000** [discordjs/discord.js#7745]
### Which package is the feature request for?

rest

### Feature

Currently, the REST manager is directly hard code to discord.js although people can change it, discord,js still install as deps and it is hard for people to separate REST manager to another process. This will be the first step to separate discord.js to a smaller service.

### Ideal solution or implementation

There are a few things to do:
- make an adapter located in each discord and REST's package where it transforms eg this.client.rest.post to a post request and decode HTTP request to rest.post.
 - change client options to DefaultRestOptions or ExternalRestOptions (URL, header, etc)
 - change init client to (something like), when init, the process will like zlib, but it first checks if the rest option is ExternalRestOptions (check is URL exists), if success it uses external rest, if failed, it tries to import rest, if success, it uses rest-client if failed it exit 1
 
 But we need to find a way to deal with DefaultRestOptions, (HTTP,rate limit, discord API) Error import from rest to discord

Extra:
- if this works we can publish a docker image for the rest manager

### Alternative solutions or implementations

While keeping discordjs easy to use (and same it current way) and allowing for microservice use, we can make a new package like discordjs/(core or bot) where it moves all current core logic of discordjs to there and the current discordjs package will like a wapper for linking all other packages, in simply (discordjs -> monolithic/wapper), (discordjs/core -> micro/bot processing part)

**Comment by H01001000** [comment]
Sidenote: can we have a roadmap for discordjs's micro package status

**Comment by kyranet** [comment]
I'm confused, what's the issue with [`@discordjs/rest`](https://www.npmjs.com/package/@discordjs/rest), which can be used standalone without the main library?

**Comment by H01001000** [comment]
TLDR completely septate rest manager out from discord.js allows:
- multiple bot processes use the same rest manager
- use other rest managers
- mockclient (in futher)

**Comment by didinele** [comment]
> mockclient (in further, where we can feed mock message directly to bot (when ws package is done) and get the response from the bot directly)

Ignoring this aspect, which I think is never going to happen - our `rest` implementation now supports proxying. We have our own `@discordjs/proxy` (and matching basic docker image!) and it's also been confirmed to work with the [twilight proxy](https://github.com/twilight-rs/http-proxy) - realistically, this addresses your first point, and makes the 2nd one unnecessary.
